### ISSUE TITLE
Can't run transaction against different cluster anymore

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#24718]
### What version of Materialize are you using?

1efa5662ef

### What is the issue?

This started failing in this range: https://github.com/MaterializeInc/materialize/compare/a87ce583b1...1efa5662ef

https://buildkite.com/materialize/nightlies/builds/6151#018d409c-ed83-4ec4-bae3-695617e5ee9b

```
^^^ +++
8:1: error: preparing query failed: db error: ERROR: current transaction is aborted, commands ignored until end of transaction block: ERROR: current transaction is aborted, commands ignored until end of transaction block
     |
   7 | 
   8 | > SELECT * FROM v1 WHERE f1 = 1;
     | ^
```
Local reproducer:
```
bin/mzcompose --find feature-benchmark down && bin/mzcompose --find feature-benchmark run default --scenario=FastPathFilterIndex
```

Smaller reproducer:
```
materialize=> CREATE MATERIALIZED VIEW v1 AS SELECT 1;
CREATE MATERIALIZED VIEW
materialize=> set cluster = quickstart;
SET
materialize=> begin;
BEGIN
materialize=*> select 1;
 ?column?
----------
        1
(1 row)

materialize=*> select * from v1;
ERROR:  querying the following items "materialize.public.v1" is not allowed from the "mz_introspection" cluster
HINT:  Use `SET CLUSTER = <cluster-name>` to change your cluster and re-run the query.
materialize=!> rollback ;
ROLLBACK
materialize=> begin;
BEGIN
materialize=*> set cluster = quickstart;
SET
materialize=*> select 1;
 ?column?
----------
        1
(1 row)

materialize=*> select * from v1;
ERROR:  querying the following items "materialize.public.v1" is not allowed from the "mz_introspection" cluster
HINT:  Use `SET CLUSTER = <cluster-name>` to change your cluster and re-run the query.
```
@mjibson I'm guessing this comes from https://github.com/MaterializeInc/materialize/pull/24628 The first `select 1` query in the transaction seems to change the cluster and there is no way to prevent that.

**Comment by def-** [comment]
I'm seeing the same failure in parallel-workload in many scenarios: https://buildkite.com/materialize/nightlies/builds/6158#_
```
materialize.data_ingest.query_error.QueryError: ('{\'S\': \'ERROR\', \'C\': \'2F003\', \'M\': \'querying the following items "db-pw-1706186397-0.s-0.t-3" is not allowed from the "mz_introspection" cluster\', \'H\': \'Use `SET CLUSTER = <cluster-name>` to change your cluster and re-run the query.\'}', 'SELECT *, mz_unsafe.mz_sleep(4) FROM "db-pw-1706186397-0"."s-0"."t-3"  LIMIT 1;')
```

**Comment by aalexandrov** [comment]
This is again something that might be fixed in `main` by #24703 because the original revert (#24677) was not sufficient.

**Comment by def-** [comment]
Works with  #24628 reverted:
```
materialize=> CREATE MATERIALIZED VIEW v1 AS SELECT 1;
CREATE MATERIALIZED VIEW
materialize=> set cluster = quickstart;
SET
materialize=> begin;
BEGIN
materialize=*> set cluster = quickstart;
SET
materialize=*> select 1;
 ?column?
----------
        1
(1 row)

materialize=*> select * from v1;
 ?column?
----------
        1
(1 row)

materialize=*> commit;
COMMIT
```

**Comment by jkosh44** [comment]
What's happening here is that the `SELECT 1` query is auto-routed to `mz_introspection`. Since it's the first query in the transaction, it locks `mz_introspection` as the cluster for the entire transaction. We aren't allowed to query user objects on `mz_introspection` so the query fails. It was technically a bug that this transaction worked previously because we don't want to allow users to query user objects on `mz_introspection`.

Though I agree this behavior is unfortunate and the error message is incorrect. You can actually prevent this by turning off autorouting `SET auto_route_introspection_queries TO false`. Another thing we can try that @mjibson previously suggested is to change clusters mid-transaction if we haven't already acquired read-holds.

**Comment by def-** [comment]
Thanks, I'll try to adapt the tests: https://github.com/MaterializeInc/materialize/pull/24725

**Comment by jkosh44** [comment]
Though, it is possible that some customer was relying on this behavior and we've accidentally broken their workflow. So we may want to think about that briefly.

**Comment by mjibson** [comment]
Those users will have to change their workflow because their queries were being running (unknown to them or us) on the introspection cluster which is very bad because those queries could easily crash that cluster causing lack of observability. I think everything is working as intended, and what we did was expose an existing bad but silent behavior.

**Comment by mjibson** [comment]
I think that error message is coming from testdrive which is retrying and only returning the last error. The first error message is helpful and correct.

**Comment by def-** [comment]
Yes, please ignore the testdrive error message for this. The tests are fixed, so if this is expected, then can be closed.

**Comment by jkosh44** [comment]
I'm talking about this error

```
ERROR:  querying the following items "materialize.public.v1" is not allowed from the "mz_introspection" cluster
HINT:  Use `SET CLUSTER = <cluster-name>` to change your cluster and re-run the query.
```

more specifically, the hint within the error. Running `SET CLUSTER` won't actually make a difference because the first query will ignore the `CLUSTER` variable and auto-route the query.

**Comment by mjibson** [comment]
Ahh, yes that's a bad hint.

Also we did some brainstorming about how to fix this generally in https://materializeinc.slack.com/archives/C02FWJ94HME/p1706286545633849 and came to the conclusion that due to upcoming pv2 work it would be at least medium difficult to have txns use multiple clusters.

**Comment by mjibson** [comment]
Should this have been closed by #24725?

**Comment by def-** [comment]
Yes, I didn't see that https://github.com/MaterializeInc/materialize/pull/24738 was also merged, so good to close.
